### PR TITLE
Adding remedy APIs to ClusterState-based policy orchestration solutions

### PR DESCRIPTION
**Pull request by XiShanYongYe-Chang**
**What type of PR is this?**

/kind feature
/kind api-change

**What this PR does / why we need it**:

Please refer to #4651 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Adding remedy APIs to ClusterCondition-based policy orchestration solutions.
```

### CODE REVIEW

**Comment by RainbowMango** [comment]
Do we have the proposal for this feature?

**Comment by XiShanYongYe-Chang** [comment]
> Do we have the proposal for this feature?

The proposal isn't ready yet.

File: charts/karmada/_crds/bases/remedy/remedy.karmada.io_clusterstates.yaml [hunk 1]
```diff
@@ -0,0 +1,69 @@
+---
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.8.0
+  creationTimestamp: null
+  name: clusterstates.remedy.karmada.io
+spec:
+  group: remedy.karmada.io
```
Comment (chaunceyjiang) [inline charts/karmada/_crds/bases/remedy/remedy.karmada.io_clusterstates.yaml:10]: What does `remedy` mean?

Comment (XiShanYongYe-Chang) [reply charts/karmada/_crds/bases/remedy/remedy.karmada.io_clusterstates.yaml:10]: The remedy is used to represent the cure after a cluster failure in Karmada. I'll submit the proposal as soon as possible.

File: pkg/apis/remedy/v1alpha1/remedy_types.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope="Cluster"
+
+// Remedy represents the cluster-level management strategies based on ClusterState.
+type Remedy struct {
```
Comment (jwcesign) [inline pkg/apis/remedy/v1alpha1/remedy_types.go:29]: How about change a name, like `RemedyPolicy`

File: pkg/apis/remedy/v1alpha1/remedy_types.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope="Cluster"
+
+// Remedy represents the cluster-level management strategies based on ClusterState.
```
Comment (RainbowMango) [inline pkg/apis/remedy/v1alpha1/remedy_types.go:28]:
```suggestion
// Remedy represents the cluster-level management strategies based on cluster conditions.
```

File: pkg/apis/cluster/types.go [hunk 4]
```diff
@@ -290,6 +290,11 @@ type ClusterStatus struct {
 	// ResourceSummary represents the summary of resources in the member cluster.
 	// +optional
 	ResourceSummary *ResourceSummary
+
+	// RemedyActions represents the remedy actions that needs to be performed
+	// on the cluster object.
```
Comment (RainbowMango) [inline pkg/apis/cluster/types.go:295]:
```suggestion
	// on the cluster.
```

File: pkg/apis/cluster/v1alpha1/types.go [hunk 5]
```diff
@@ -302,6 +302,11 @@ type ClusterStatus struct {
 	// ResourceSummary represents the summary of resources in the member cluster.
 	// +optional
 	ResourceSummary *ResourceSummary `json:"resourceSummary,omitempty"`
+
+	// RemedyActions represents the remedy actions that needs to be performed
+	// on the cluster object.
```
Comment (RainbowMango) [inline pkg/apis/cluster/v1alpha1/types.go:307]: ditto.

File: pkg/apis/remedy/v1alpha1/remedy_types.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	// For clusters that meet the DecisionConditions, Actions will be preformed.
+	// If empty, all clusters will be selected.
+	// +optional
+	ClusterAffinity *ClusterAffinity `json:"clusterAffinity,omitempty"`
+
+	// DecisionControls indicates the decision matches of triggering
+	// the remedy system to perform the actions.
+	// As long as any one DecisionControl matches, the Actions will be preformed.
+	// If empty, the Actions will be performed immediately.
+	// +optional
```
Comment (RainbowMango) [inline pkg/apis/remedy/v1alpha1/remedy_types.go:50]: `DecisionControls`?

**Comment by XiShanYongYe-Chang** [comment]
Hi @RainbowMango all comments have been updated.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**